[FLINK-8703][tests] Port KafkaTestBase to MiniClusterResource

zentol opened the pull request:
## What is the purpose of the change

Ports the `KafkaTestBase` and extending classes to use `MiniClusterResource`.

## Verifying this change

Run all tests extending `KafkaTestBase` with `flip6` profile enabled/disabled.

aljoscha commented on flink-connectors/flink-connector-kafka-0.10/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka010ITCase.java:
```diff
@@ -81,12 +81,6 @@ public void testCancelingEmptyTopic() throws Exception {
 	public void testCancelingFullTopic() throws Exception {
 		runCancelingOnFullInputTest();
 	}
-
```
Why are those removed?

zentol replied:
ah right, forgot to comment that. I found these test to be rather..._odd_. They check behavior when not enough slots are available, but in the old code afaik this fails before the client even submits the job, and in flip6 this stalls as we never check whether enough slots are available (i guess with the underlying assumption that we would just allocate more TMs until we have enough).

pnowojski commented on flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -244,7 +255,7 @@ public void run() {
 
 		// cancel the job & wait for the job to finish
-		JobManagerCommunicationUtils.cancelCurrentJob(flink.getLeaderGateway(timeout));
+		client.cancel(getRunningJobs(client).get(0));
```
`Iterables.getOnlyElement(getRunningJobs(client))`?

pnowojski commented on flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -132,8 +140,11 @@
-		JobManagerCommunicationUtils.waitUntilNoJobIsRunning(flink.getLeaderGateway(timeout));
+	public void setClientAndEnsureNoJobIsLingering() throws Exception {
+		client = flink.getClusterClient();
+		while (!getRunningJobs(client).isEmpty()){
```
```
		while (!getRunningJobs(client).isEmpty()){
			Thread.sleep(50);
		}
```

This is being copied pasted couple of times. Please extract to common method. Maybe to an equivalent of `JobManagerCommunicationUtils` or even to some `TestingClusterClient`?

pnowojski commented on flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1063,23 +1078,27 @@ public void runCancelingOnEmptyInputTest() throws Exception {
 
 		final AtomicReference<Throwable> error = new AtomicReference<>();
 
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
```
Please unify and extract common code between `runCancelingOnFullInputTest` and `runCancelingOnEmptyInputTest` for example into:

```
	private void runCancelingTest(boolean emptyInput) throws Exception {
		final String topic = emptyInput ? "cancelingOnEmptyInputTopic" : "cancelingOnFullTopic";

		final int parallelism = 3;
		createTestTopic(topic, parallelism, 1);

		// launch a producer thread
		DataGenerators.InfiniteStringsGenerator generator =
			new DataGenerators.InfiniteStringsGenerator(kafkaServer, topic);
		if (!emptyInput) {
			generator.start();
		}

		// launch a consumer asynchronously

		final AtomicReference<Throwable> jobError = new AtomicReference<>();

		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
		env.setParallelism(parallelism);
		env.enableCheckpointing(100);
		env.getConfig().disableSysoutLogging();

		Properties props = new Properties();
		props.putAll(standardProps);
		props.putAll(secureProps);
		FlinkKafkaConsumerBase<String> source = kafkaServer.getConsumer(topic, new SimpleStringSchema(), props);

		env.addSource(source).addSink(new DiscardingSink<String>());

		JobGraph jobGraph = StreamingJobGraphGenerator.createJobGraph(env.getStreamGraph());
		final JobID jobId = jobGraph.getJobID();

		final Runnable jobRunner = new Runnable() {
			@Override
			public void run() {
				try {
					client.setDetached(false);
					client.submitJob(jobGraph, KafkaConsumerTestBase.class.getClassLoader());
				}
				catch (Throwable t) {
					jobError.set(t);
				}
			}
		};

		Thread runnerThread = new Thread(jobRunner, "program runner thread");
		runnerThread.start();

		// wait a bit before canceling
		Thread.sleep(2000);

		Throwable failueCause = jobError.get();
		if (failueCause != null) {
			failueCause.printStackTrace();
			Assert.fail("Test failed prematurely with: " + failueCause.getMessage());
		}

		// cancel
		client.cancel(jobId);

		// wait for the program to be done and validate that we failed with the right exception
		runnerThread.join();

		assertEquals(JobStatus.CANCELED, client.getJobStatus(jobId).get());

		if (generator.isAlive()) {
			generator.shutdown();
			generator.join();
		}
		else if (!emptyInput) {
			Throwable t = generator.getError();
			if (t != null) {
				t.printStackTrace();
				fail("Generator failed: " + t.getMessage());
			} else {
				fail("Generator failed with no exception");
			}
		}

		deleteTestTopic(topic);
	}
```

zentol replied:
That would go further than porting the test; I'd rather not do that in this PR.

pnowojski replied:
As you wish, I can open a follow up since it's a trivial fixup.

zentol commented:
I've rebased and updated the PR.

The `JobManagerCommunicationUtils` were reworked into a `ClusterCommunicationUtils`. All contained methods were now unused. We are also using `Iterables.getOnlyElement` now,

zentol commented:
merging.
